### ISSUE TITLE
Slow http2 framer when under heavy load

### ISSUE THREAD

**Issue by carl-mastrangelo** [netty/netty#5360]
I have had this on my plate for a while but it is getting too far behind and I dont want to forget it.  

When sending a lot of http requests, there are a lot of tiny buffers created for the http2 frame headers.  While allocation is fast, writing them is very slow when under load.  Writes to the network eventually end up as syscalls to writev, but that can only write at most 1024 (IOV_MAX) buffers per syscall.  This is not normally a problem, except when sending lots of tiny buffers.  Whats worse is that writev takes a vector of buffers, but returns the number of written bytes, so the entire list of buffers has to be iterated through again when building the vector for the next syscall.  Netty by default gives up after 16 iterations of this and just waits until the next chance to write the rest of the values.  

The effects of this are not noticeable under it starts falling behind. Once it does fall behind, memory usage starts going up fast.  Not all http2 frames count against the flow control window, so memory will grow unbounded (the max concurrent streams  limit helps, but that may not be turned on).  In heap dump analysis, there were hundreds of thousands of 4, 13, and 14 byte buffers, corresponding to WINDOW_SIZE and SETTINGS frames (and maybe HEADERS).

I had a semi working demo where writes to the DefaultHttp2Framer would just buffer memory writes into a small (64-128 sized ByteBuf) and just flush it when it filled up or an explicit flush happened.  In my tests, the program would keep up with the load for about 30 seconds, and then memory would sky rocket until the GC kicks in, and finally the entire program would spend 100% trying to free the couple thousand bytebuf's that were written.

**Comment by buchgr** [comment]
Did this happen under a real workload?

**Comment by carl-mastrangelo** [comment]
Yes, it happens under the same load that caused https://github.com/netty/netty/issues/5186 .  The characteristics are very similar (slow down to a crawl, basically get nothing done).  The example code (which I cannot find) that was used to find that bug is the same one that causes this.   basically just make a ton of small http2 calls as fast as you can to force the problem. (over a local tcp link)

**Comment by Scottmitch** [comment]
@carl-mastrangelo - Is this the repro discussed here https://github.com/netty/netty/pull/5191#issuecomment-215952785 (https://github.com/carl-mastrangelo/grpc-java/commit/786a96f6bf14035ca81849285ad5a65a4fe2dd91) what you are looking for?

**Comment by carl-mastrangelo** [comment]
@Scottmitch  yes!  Removing the cancellation makes the bug happen much more quickly.  Whether or not that is a realistic workload is debateable, but it is basically what our load tests/ benchmarks do.  

On a side note, while profiling our benchmarks, the majority of the time is spent in the `writev` syscall, which I attribute to writing too many small buffers.  Even if Netty doesn't get backed up, it's still slow.

**Comment by Scottmitch** [comment]
This issue seems related to https://github.com/netty/netty/issues/3998 but for egress.

**Comment by buchgr** [comment]
I ll get on it :-)

**Comment by Scottmitch** [comment]
@buchgr - great! what is your plan to mitigate this issue?

**Comment by buchgr** [comment]
@Scottmitch don't know yet ... thankful for any suggestions! :-)

@carl-mastrangelo 
> heap dump analysis, there were hundreds of thousands of 4, 13, and 14 byte buffers, corresponding to WINDOW_SIZE and SETTINGS frames (and maybe HEADERS).

Hmm ... hundreds of thousands `WINDOW_SIZE` and `SETTINGS`? I would expect all of that to be `HEADERS` frames?

I also ran that code, removed cancel and the semaphore and found that ~2/3 of the live heap are command objects in grpc's `WriteQueue` and 1/3 is the `ChannelOutboundBuffer` totaling about 4 million objects. Looking at a heap dump, I noticed that the cached thread-local [`ByteBuffer` array](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java#L379) is always at it's initial size of 1024, but I would have expected that to be [much bigger](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java#L408) with such a backlog. Turns out, in grpc we are flushing so much that we only ever write [2 or 4 buffers at a time](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java#L381).

So it seems to me that in this case the problem is not that we are passing too many small buffers to `writev`, but calling `writev` too often with too few buffers.

Any thoughts @carl-mastrangelo @ejona86 @nmittler ?

**Comment by ejona86** [comment]
gRPC's [WriteQueue](https://github.com/grpc/grpc-java/blob/v0.14.0/netty/src/main/java/io/grpc/netty/WriteQueue.java) should auto-coalesce writes when under load. The cancel in that test was important for saturating the channel outbound, otherwise the bottleneck may be due to latency and outbound will be consistently drained.

How is it possible for over a million objects to be in `ChannelOutboundBuffer` but "only ever write 2 or 4 buffers at a time"?

**Comment by buchgr** [comment]
@ejona86 

> gRPC's WriteQueue should auto-coalesce writes when under load. The cancel in that test was important for saturating the channel outbound, otherwise the bottleneck may be due to latency and outbound will be consistently drained.

I modified the test to just create new calls in an endless loop.

> How is it possible for over a million objects to be in ChannelOutboundBuffer but "only ever write 2 or 4 buffers at a time"?

that's what I don't understand 😅

**Comment by normanmaurer** [comment]
@buchgr are you really sure about the 2 / 4 buffers at a time ? addFlush() should mark everything ready to be flushed that exists in the `ChannelOutboundBuffer`. The only way it could end up writing less would be if writev would do a partial write and so the channel becomes unwritable. Then the channel becomes writable again and we try to pass all buffers again to writev which again does a partial write. This will continue until everything is written. But this would mean your network stack is saturated /cc @ejona86

**Comment by buchgr** [comment]
@ejona86 ohhh I am an idiot. I ran an outdated jar ...

**Comment by buchgr** [comment]
I investigated a bit more. I added code to track the max number of buffers passed to `SocketChannel.write`, as well as logging to find when and why a write fails. Running the `NoiseClient`, we at most pass 256 buffers to `SocketChannel.write` (limited by grpc's write queue). We fail cause, `SocketChannel.write` returns `0` (it returns the number of bytes written). It's then when the `ChannelOutboundBuffer` starts to grow. I assume it's because we are filling up the send buffer. So it still seems to be true that we are not hitting `IOV_MAX` or the write spin count. This is a grpc issue though and being addressed in https://github.com/grpc/grpc-java/pull/1989.
